Atom will not build against iojs 3.2.0+ / node 4

I have commented on the node-runas issue which I believe to be the cause of this, but perhaps it should be a new issue (and will get more exposure here):

```
../node_modules/nan/nan.h: In function ‘v8::Local<v8::Object> NanBufferUse(char*, uint32_t)’:
../node_modules/nan/nan.h:682:12: error: ‘Use’ is not a member of ‘node::Buffer’
     return node::Buffer::Use(v8::Isolate::GetCurrent(), data, size);
            ^
make: *** [Release/obj.target/runas/src/main.o] Error 1
gyp ERR! build error 
gyp ERR! stack Error: `make` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (/usr/lib64/node_modules/npm/node_modules/node-gyp/lib/build.js:269:23)
gyp ERR! stack     at emitTwo (events.js:87:13)
gyp ERR! stack     at ChildProcess.emit (events.js:172:7)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:200:12)
gyp ERR! System Linux 3.10.17
gyp ERR! command "/usr/bin/iojs" "/usr/lib64/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
gyp ERR! cwd /tmp/SBo/atom-1.0.7/build/node_modules/runas
gyp ERR! node -v v3.2.0
gyp ERR! node-gyp -v v2.0.2
gyp ERR! not ok 
npm ERR! Linux 3.10.17
npm ERR! argv "/usr/bin/iojs" "/usr/bin/npm" "--userconfig=/tmp/SBo/atom-1.0.7/build/.npmrc" "install" "--loglevel" "error"
npm ERR! node v3.2.0
npm ERR! npm  v2.13.3
npm ERR! code ELIFECYCLE

npm ERR! runas@2.0.0 install: `node-gyp rebuild`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the runas@2.0.0 install script 'node-gyp rebuild'.
npm ERR! This is most likely a problem with the runas package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node-gyp rebuild
npm ERR! You can get their info via:
npm ERR!     npm owner ls runas
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /tmp/SBo/atom-1.0.7/build/npm-debug.log
```

Here is the full npm-debug.log: https://gist.github.com/ryanpcmcquen/8e798b2d172b96c5d9ab

And the node-runas issue:

https://github.com/atom/node-runas/issues/11

Idem for me. With master branch. 
Node: v4.1.1
npm: v2.14.4
Installing build modules...
In file included from ../src/main.cc:1:0:
../node_modules/nan/nan.h:260:25: error: redefinition of ‘template<class T> v8::Local<T> _NanEnsureLocal(v8::Local<T>)’
 NAN_INLINE v8::Local<T> _NanEnsureLocal(v8::Local<T> val) {
                         ^
../node_modules/nan/nan.h:255:25: note: ‘template<class T> v8::Local<T> _NanEnsureLocal(v8::Handle<T>)’ previously declared here
 NAN_INLINE v8::Local<T> _NanEnsureLocal(v8::Handle<T> val) {
                         ^
../node_modules/nan/nan.h:660:13: error: ‘node::smalloc’ has not been declared
     , node::smalloc::FreeCallback callback
             ^
../node_modules/nan/nan.h:660:35: error: expected ‘,’ or ‘...’ before ‘callback’
     , node::smalloc::FreeCallback callback
                                   ^
../node_modules/nan/nan.h: In function ‘v8::Local<v8::Object> NanNewBufferHandle(char*, size_t, int)’:
../node_modules/nan/nan.h:664:50: error: ‘callback’ was not declared in this scope
         v8::Isolate::GetCurrent(), data, length, callback, hint);
                                                  ^
../node_modules/nan/nan.h:664:60: error: ‘hint’ was not declared in this scope
         v8::Isolate::GetCurrent(), data, length, callback, hint);
                                                            ^
../node_modules/nan/nan.h: In function ‘v8::Local<v8::Object> NanNewBufferHandle(const char*, uint32_t)’:
../node_modules/nan/nan.h:671:67: error: call of overloaded ‘New(v8::Isolate*, const char*&, uint32_t&)’ is ambiguous
     return node::Buffer::New(v8::Isolate::GetCurrent(), data, size);
                                                                   ^
../node_modules/nan/nan.h:671:67: note: candidates are:
In file included from ../node_modules/nan/nan.h:25:0,
                 from ../src/main.cc:1:
/root/.node-gyp/4.1.1/include/node/node_buffer.h:31:40: note: v8::MaybeLocal<v8::Object> node::Buffer::New(v8::Isolate*, v8::Local<v8::String>, node::encoding) <near match>
 NODE_EXTERN v8::MaybeLocal<v8::Object> New(v8::Isolate* isolate,
                                        ^
/root/.node-gyp/4.1.1/include/node/node_buffer.h:31:40: note:   no known conversion for argument 3 from ‘uint32_t {aka unsigned int}’ to ‘node::encoding’
/root/.node-gyp/4.1.1/include/node/node_buffer.h:43:40: note: v8::MaybeLocal<v8::Object> node::Buffer::New(v8::Isolate*, char*, size_t) <near match>
 NODE_EXTERN v8::MaybeLocal<v8::Object> New(v8::Isolate* isolate,
                                        ^
/root/.node-gyp/4.1.1/include/node/node_buffer.h:43:40: note:   no known conversion for argument 2 from ‘const char*’ to ‘char*’
In file included from ../src/main.cc:1:0:
../node_modules/nan/nan.h: In function ‘v8::Local<v8::Object> NanNewBufferHandle(uint32_t)’:
../node_modules/nan/nan.h:675:61: error: could not convert ‘node::Buffer::New(v8::Isolate::GetCurrent(), ((size_t)size))’ from ‘v8::MaybeLocal<v8::Object>’ to ‘v8::Local<v8::Object>’
     return node::Buffer::New(v8::Isolate::GetCurrent(), size);
                                                             ^
../node_modules/nan/nan.h: In function ‘v8::Local<v8::Object> NanBufferUse(char*, uint32_t)’:
../node_modules/nan/nan.h:682:12: error: ‘Use’ is not a member of ‘node::Buffer’
     return node::Buffer::Use(v8::Isolate::GetCurrent(), data, size);
            ^
make: *** [Release/obj.target/runas/src/main.o] Errore 1
gyp ERR! build error 
gyp ERR! stack Error: `make` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (/usr/lib64/node_modules/npm/node_modules/node-gyp/lib/build.js:270:23)
gyp ERR! stack     at emitTwo (events.js:87:13)
gyp ERR! stack     at ChildProcess.emit (events.js:172:7)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:200:12)
gyp ERR! System Linux 4.2.1-cf
gyp ERR! command "/usr/bin/node" "/usr/lib64/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
gyp ERR! cwd /tmp/pkg/atom-master/build/node_modules/runas
gyp ERR! node -v v4.1.1
gyp ERR! node-gyp -v v3.0.3
gyp ERR! not ok 
npm ERR! Linux 4.2.1-cf
npm ERR! argv "/usr/bin/node" "/usr/bin/npm" "--userconfig=/tmp/pkg/atom-master/build/.npmrc" "install" "--loglevel" "error"
npm ERR! node v4.1.1
npm ERR! npm  v2.14.4
npm ERR! code ELIFECYCLE

npm ERR! runas@2.0.0 install: `node-gyp rebuild`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the runas@2.0.0 install script 'node-gyp rebuild'.
npm ERR! This is most likely a problem with the runas package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node-gyp rebuild
npm ERR! You can get their info via:
npm ERR!     npm owner ls runas
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /tmp/pkg/atom-master/build/npm-debug.log


Error also cz-electron-32 branch

--- a/package.json
+++ b/package.json
@@ -12,7 +12,7 @@
     "url": "https://github.com/atom/atom/issues"
   },
   "license": "MIT",
-  "electronVersion": "0.33.1",
+  "electronVersion": "0.33.3",
   "dependencies": {
     "async": "0.2.6",
     "atom-keymap": "^6.1.0",

--- a/package.json
+++ b/package.json
@@ -12,7 +12,7 @@
     "url": "https://github.com/atom/atom/issues"
   },
   "license": "MIT",
-  "electronVersion": "0.33.5",
+  "electronVersion": "0.33.6",
   "dependencies": {
     "async": "0.2.6",
     "atom-keymap": "^6.1.0",

--- a/package.json
+++ b/package.json
@@ -12,7 +12,7 @@
     "url": "https://github.com/atom/atom/issues"
   },
   "license": "MIT",
-  "electronVersion": "0.33.3",
+  "electronVersion": "0.33.4",
   "dependencies": {
     "async": "0.2.6",
     "atom-keymap": "^6.1.0",

--- a/package.json
+++ b/package.json
@@ -115,7 +115,7 @@
     "update-package-dependencies": "0.10.0",
     "welcome": "0.31.0",
     "whitespace": "0.31.0",
-    "wrap-guide": "0.38.0",
+    "wrap-guide": "0.38.1",
     "language-c": "0.49.0",
     "language-clojure": "0.18.0",
     "language-coffee-script": "0.42.0",

--- a/package.json
+++ b/package.json
@@ -12,7 +12,7 @@
     "url": "https://github.com/atom/atom/issues"
   },
   "license": "MIT",
-  "electronVersion": "0.30.7",
+  "electronVersion": "0.33.1",
   "dependencies": {
     "async": "0.2.6",
     "atom-keymap": "^6.1.0",

--- a/package.json
+++ b/package.json
@@ -12,7 +12,7 @@
     "url": "https://github.com/atom/atom/issues"
   },
   "license": "MIT",
-  "electronVersion": "0.33.4",
+  "electronVersion": "0.33.5",
   "dependencies": {
     "async": "0.2.6",
     "atom-keymap": "^6.1.0",

--- a/spec/compile-cache-spec.coffee
+++ b/spec/compile-cache-spec.coffee
@@ -18,7 +18,7 @@ describe 'CompileCache', ->
     CompileCache.resetCacheStats()
 
     spyOn(Babel, 'transform').andReturn {code: 'the-babel-code'}
-    spyOn(CoffeeScript, 'compile').andReturn {js: 'the-coffee-code', v3SourceMap: {}}
+    spyOn(CoffeeScript, 'compile').andReturn {js: 'the-coffee-code', v3SourceMap: "{}"}
     spyOn(TypeScriptSimple::, 'compile').andReturn 'the-typescript-code'
     spyOn(CSONParser, 'parse').andReturn {the: 'cson-data'}
 

--- a/spec/fixtures/babel/invalid.js
+++ b/spec/fixtures/babel/invalid.js
@@ -1,3 +1,3 @@
 'use 6to6';
 
-module.exports = v => v + 1
+module.exports = async function hello() {}

--- a/spec/lines-yardstick-spec.coffee
+++ b/spec/lines-yardstick-spec.coffee
@@ -76,10 +76,10 @@ describe "LinesYardstick", ->
 
       expect(linesYardstick.pixelPositionForScreenPosition([0, 0])).toEqual({left: 0, top: 0})
       expect(linesYardstick.pixelPositionForScreenPosition([0, 1])).toEqual({left: 7, top: 0})
-      expect(linesYardstick.pixelPositionForScreenPosition([0, 5])).toEqual({left: 37.8046875, top: 0})
-      expect(linesYardstick.pixelPositionForScreenPosition([1, 6])).toEqual({left: 43.20703125, top: 14})
-      expect(linesYardstick.pixelPositionForScreenPosition([1, 9])).toEqual({left: 72.20703125, top: 14})
-      expect(linesYardstick.pixelPositionForScreenPosition([2, Infinity])).toEqual({left: 288.046875, top: 28})
+      expect(linesYardstick.pixelPositionForScreenPosition([0, 5])).toEqual({left: 37.78125, top: 0})
+      expect(linesYardstick.pixelPositionForScreenPosition([1, 6])).toEqual({left: 43.171875, top: 14})
+      expect(linesYardstick.pixelPositionForScreenPosition([1, 9])).toEqual({left: 72.171875, top: 14})
+      expect(linesYardstick.pixelPositionForScreenPosition([2, Infinity])).toEqual({left: 287.859375, top: 28})
 
     it "reuses already computed pixel positions unless it is invalidated", ->
       atom.styles.addStyleSheet """
@@ -105,9 +105,9 @@ describe "LinesYardstick", ->
 
       linesYardstick.invalidateCache()
 
-      expect(linesYardstick.pixelPositionForScreenPosition([1, 2])).toEqual({left: 24.00390625, top: 14})
-      expect(linesYardstick.pixelPositionForScreenPosition([2, 6])).toEqual({left: 72.01171875, top: 28})
-      expect(linesYardstick.pixelPositionForScreenPosition([5, 10])).toEqual({left: 120.01171875, top: 70})
+      expect(linesYardstick.pixelPositionForScreenPosition([1, 2])).toEqual({left: 24, top: 14})
+      expect(linesYardstick.pixelPositionForScreenPosition([2, 6])).toEqual({left: 72, top: 28})
+      expect(linesYardstick.pixelPositionForScreenPosition([5, 10])).toEqual({left: 120, top: 70})
 
     it "correctly handles RTL characters", ->
       atom.styles.addStyleSheet """
@@ -156,7 +156,7 @@ describe "LinesYardstick", ->
         expect(linesYardstick.screenPositionForPixelPosition({top: 32, left: 24.3})).toEqual([2, 3])
         expect(linesYardstick.screenPositionForPixelPosition({top: 46, left: 66.5})).toEqual([3, 9])
         expect(linesYardstick.screenPositionForPixelPosition({top: 80, left: 99.9})).toEqual([5, 14])
-        expect(linesYardstick.screenPositionForPixelPosition({top: 80, left: 224.4365234375})).toEqual([5, 29])
+        expect(linesYardstick.screenPositionForPixelPosition({top: 80, left: 224.2365234375})).toEqual([5, 29])
         expect(linesYardstick.screenPositionForPixelPosition({top: 80, left: 225})).toEqual([5, 30])
 
       it "clips pixel positions above buffer start", ->

--- a/spec/module-cache-spec.coffee
+++ b/spec/module-cache-spec.coffee
@@ -86,6 +86,7 @@ describe 'ModuleCache', ->
       exports.load = function() { require('underscore-plus'); };
     """
 
+    spyOn(process, 'cwd').andReturn('/') # Required when running this test from CLI
     packageMain = require(indexPath)
     Module._findPath.reset()
     expect(-> packageMain.load()).toThrow()

--- a/spec/text-editor-component-spec.coffee
+++ b/spec/text-editor-component-spec.coffee
@@ -3482,12 +3482,14 @@ describe "TextEditorComponent", ->
 
         editor.moveRight()
         nextAnimationFrame()
-        right = wrapperNode.pixelPositionForScreenPosition([0, 6]).left
+
+        margin = component.presenter.getHorizontalScrollMarginInPixels()
+        right = wrapperNode.pixelPositionForScreenPosition([0, 4]).left + margin
         expect(wrapperNode.getScrollRight()).toBeCloseTo right, 0
 
         editor.moveRight()
         nextAnimationFrame()
-        right = wrapperNode.pixelPositionForScreenPosition([0, 7]).left
+        right = wrapperNode.pixelPositionForScreenPosition([0, 5]).left + margin
         expect(wrapperNode.getScrollRight()).toBeCloseTo right, 0
 
       it "scrolls left when the last cursor gets closer than ::horizontalScrollMargin to the left of the editor", ->
@@ -3499,12 +3501,14 @@ describe "TextEditorComponent", ->
 
         editor.moveLeft()
         nextAnimationFrame()
-        left = wrapperNode.pixelPositionForScreenPosition([6, 59]).left
+
+        margin = component.presenter.getHorizontalScrollMarginInPixels()
+        left = wrapperNode.pixelPositionForScreenPosition([6, 61]).left - margin
         expect(wrapperNode.getScrollLeft()).toBeCloseTo left, 0
 
         editor.moveLeft()
         nextAnimationFrame()
-        left = wrapperNode.pixelPositionForScreenPosition([6, 58]).left
+        left = wrapperNode.pixelPositionForScreenPosition([6, 60]).left - margin
         expect(wrapperNode.getScrollLeft()).toBeCloseTo left, 0
 
       it "scrolls down when inserting lines makes the document longer than the editor's height", ->

--- a/src/application-delegate.coffee
+++ b/src/application-delegate.coffee
@@ -2,6 +2,7 @@ _ = require 'underscore-plus'
 ipc = require 'ipc'
 remote = require 'remote'
 shell = require 'shell'
+webFrame = require 'web-frame'
 {Disposable} = require 'event-kit'
 {getWindowLoadSettings, setWindowLoadSettings} = require './window-load-settings-helpers'
 
@@ -163,3 +164,6 @@ class ApplicationDelegate
 
   openExternal: (url) ->
     shell.openExternal(url)
+
+  disablePinchToZoom: ->
+    webFrame.setZoomLevelLimits(1, 1)

--- a/src/atom-environment.coffee
+++ b/src/atom-environment.coffee
@@ -1,5 +1,6 @@
 crypto = require 'crypto'
 path = require 'path'
+webFrame = require 'web-frame'
 
 _ = require 'underscore-plus'
 {deprecate} = require 'grim'
@@ -187,6 +188,9 @@ class AtomEnvironment extends Model
 
     @stylesElement = @styles.buildStylesElement()
     @document.head.appendChild(@stylesElement)
+
+    # Disable pinch-to-zoom.
+    webFrame.setZoomLevelLimits(1, 1)
 
     @keymaps.subscribeToFileReadFailure()
     @keymaps.loadBundledKeymaps()

--- a/src/atom-environment.coffee
+++ b/src/atom-environment.coffee
@@ -1,6 +1,5 @@
 crypto = require 'crypto'
 path = require 'path'
-webFrame = require 'web-frame'
 
 _ = require 'underscore-plus'
 {deprecate} = require 'grim'
@@ -189,8 +188,7 @@ class AtomEnvironment extends Model
     @stylesElement = @styles.buildStylesElement()
     @document.head.appendChild(@stylesElement)
 
-    # Disable pinch-to-zoom.
-    webFrame.setZoomLevelLimits(1, 1)
+    @applicationDelegate.disablePinchToZoom()
 
     @keymaps.subscribeToFileReadFailure()
     @keymaps.loadBundledKeymaps()

--- a/src/browser/atom-protocol-handler.coffee
+++ b/src/browser/atom-protocol-handler.coffee
@@ -29,7 +29,7 @@ class AtomProtocolHandler
 
   # Creates the 'atom' custom protocol handler.
   registerAtomProtocol: ->
-    protocol.registerProtocol 'atom', (request) =>
+    protocol.registerFileProtocol 'atom', (request, callback) =>
       relativePath = path.normalize(request.url.substr(7))
 
       if relativePath.indexOf('assets/') is 0
@@ -41,4 +41,4 @@ class AtomProtocolHandler
           filePath = path.join(loadPath, relativePath)
           break if fs.statSyncNoException(filePath).isFile?()
 
-      new protocol.RequestFileJob(filePath)
+      callback(filePath)

--- a/src/browser/atom-window.coffee
+++ b/src/browser/atom-window.coffee
@@ -88,7 +88,7 @@ class AtomWindow
       hash: encodeURIComponent(JSON.stringify(loadSettings))
 
   getLoadSettings: ->
-    if @browserWindow.webContents?.loaded
+    if @browserWindow.webContents? and not @browserWindow.webContents.isLoading()
       hash = url.parse(@browserWindow.webContents.getUrl()).hash.substr(1)
       JSON.parse(decodeURIComponent(hash))
 

